app/controllers/spree/products_controller_decorator.rb

```diff
@@ -1,44 +1,56 @@
 Spree::ProductsController.class_eval do
+  before_action :load_data, only: :index
   before_action(only: :index) do |controller|
     if controller.request.format.rss?
-      load_feed_products
+      feed_products
     end
   end
 
+  respond_to :html, :rss
+
   def index
     respond_to do |format|
-      if product_feed
-        format.rss { render inline: xml.generate, layout: false }
-      else
-        format.rss { redirect_to action: 'index' }
-      end
-
+      format.rss { render inline: xml.generate, layout: false }
       format.html
     end
   end
 
   private
 
-  def load_feed_products
+  def load_data
+    @searcher = build_searcher(params.merge(include_images: true))
+    @products = @searcher.retrieve_products
+    @taxonomies = Spree::Taxonomy.includes(root: :children)
+  end
+
+  def product_feed
+    @product_feed ||= Spree::ProductFeed.
+                      where('lower(name) = lower(?)', params[:feed]).first
+  end
+
+  def product_catalog
+    @product_catalog ||= product_feed.try(:product_catalog)
+  end
+
+  def feed_products
+    @feed_products ||= load_items
+  end
+
+  def load_items
     items = []
-    product_catalog = product_feed.try(:product_catalog)
 
     Spree::Variant.where(id: product_catalog.item_ids).each do |variant|
```

Comment:
I think we talked about this section during the first iteration of this feature. It's something we will have to address soon, as we have a client starting this quarter whose product catalog is several thousand products (each with multiple variants), so their catalog would break this controller action for sure.